### PR TITLE
User controller testing

### PR DESCRIPTION
**Pull request by crod951**

### CODE REVIEW

File: test/epochtalk_server_web/controllers/user_controller_test.exs [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+  @invalid_username_attrs %{username: "", email: "invalidtest@test.com", password: "password"}
+  @invalid_password_attrs %{username: "invalid", email: "invalidtest@test.com", password: ""}
+
+  @login_create_attrs %{username: "logintest", email: "logintest@test.com", password: "password"}
+  @login_attrs %{username: "logintest", password: "password"}
+  @invalid_login_password_attrs %{username: "logintest", password: "1"}
+  @invalid_login_username_attrs %{username: "test", password: "password"}
+
+
+  describe "check username and email" do
```
Comment (unenglishable) [inline test/epochtalk_server_web/controllers/user_controller_test.exs:29]: have route names/paths or method names in describe

what parts of the project are we testing?
perhaps split out tests that don't fall under the same method or route

File: test/epochtalk_server_web/controllers/user_controller_test.exs [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    setup [:create_user]
+
+    test "check if username is taken", %{conn: conn, user: user} do
```
Comment (unenglishable) [inline test/epochtalk_server_web/controllers/user_controller_test.exs:32]: in test name, describe the action and desired outcome

what are we checking for, and what should the code do given particular input?

File: test/epochtalk_server_web/controllers/user_controller_test.exs [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+      conn = get(conn, Routes.user_path(conn, :username, user.username))
+      assert %{"found" => true} = json_response(conn, 200)
+    end
+
+    test "check if email is taken", %{conn: conn, user: user} do
+      conn = get(conn, Routes.user_path(conn, :email, user.email))
+      assert %{"found" => true} = json_response(conn, 200)
+    end
+  end
```
Comment (unenglishable) [inline test/epochtalk_server_web/controllers/user_controller_test.exs:41]: a solid test suite will check all code paths to ensure that the code works as expected in each case

consider the different possible starting conditions and what the results should be

also consider if there are any edge cases and test those as well

these suggestions apply to all tests

File: test/epochtalk_server_web/controllers/user_controller_test.exs (hunk 1 above)
Comment (unenglishable) [reply test/epochtalk_server_web/controllers/user_controller_test.exs:29]: checking out [this blog](https://brooklinmyers.medium.com/phoenix-elixir-testing-beyond-mix-test-5b07de241001)...  i think the descriptions in this example are pretty useful.

example

here, they note the tested functions and desired outcome.  this one shows the function with arity `(/2)`, which specifies exactly which version of the function they're testing.

File: test/epochtalk_server_web/controllers/user_controller_test.exs (hunk 3 above)
Comment (unenglishable) [reply test/epochtalk_server_web/controllers/user_controller_test.exs:41]: for example:
in this test, we're checking email and username routes.  your test checks the functionality if username and email are already taken; it should also check if the functions return the right result when they aren't taken as well.

File: test/epochtalk_server_web/controllers/user_controller_test.exs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+  describe "authenticate" do
+    @tag :authenticated
+    test "authenticate success", %{conn: conn} do
+      conn = get(conn, Routes.user_path(conn, :authenticate))
+      {:ok, user} = User.by_username(@auth_attrs.username)
+      assert user.id == String.to_integer(json_response(conn, 200)["id"])
+    end
+
+    test "renders error when user is not loegged in", %{conn: conn} do
```
Comment (unenglishable) [inline test/epochtalk_server_web/controllers/user_controller_test.exs:248]: * logged in

File: test/epochtalk_server_web/controllers/user_controller_test.exs [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    end
+
+    test "renders errors when password is missing", %{conn: conn} do
+      conn = post(conn, Routes.user_path(conn, :register, @invalid_password_attrs))
+      assert %{ "error" => "Bad Request",
+                "message" => "Password can't be blank"} = json_response(conn, 400)
+    end
+  end
+
+  describe "confirm" do
```
Comment (unenglishable) [inline test/epochtalk_server_web/controllers/user_controller_test.exs:118]: nice work here, we can tell you read the code and know how it works 👍

**Review by unenglishable — CHANGES_REQUESTED** [review]
looks good so far, just some minor improvements to be made;

extra cases for checking fail/success to check more code paths

better naming will help us understand test failures at a glance, letting us know what's being tested and what the expected result is.  some of this is revealed by how elixir handles returning test results (says which line errored and what the expected result was), but it's also nice to have the descriptions match up

after these quick fixes, we'll merge in and talk about the next steps in a meet.  nice work!
